File: src/main/java/net/onelitefeather/pandorascluster/land/position/HomePosition.java
package net.onelitefeather.pandorascluster.land.position;

import jakarta.persistence.*;
import net.onelitefeather.pandorascluster.land.Land;
import net.onelitefeather.pandorascluster.land.player.LandPlayer;
import org.bukkit.Bukkit;
import org.bukkit.Location;
import org.bukkit.World;
import org.jetbrains.annotations.NotNull;

@Entity
public class HomePosition {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column
    private double x;

    @Column
    private double y;

    @Column
    private double z;

    @Column
    private float yaw;

    @Column
    private float pitch;

    @OneToOne
    private Land land;

    public HomePosition() {
    }

    public HomePosition(double x, double y, double z, float yaw, float pitch) {
        this.x = x;
        this.y = y;
        this.z = z;
        this.yaw = yaw;
        this.pitch = pitch;
    }

    public Long getId() {
        return id;
    }

    public void setId(Long id) {
        this.id = id;
    }

    public double getX() {
        return x;
    }

    public void setX(double x) {
        this.x = x;
    }

    public double getY() {
        return y;
    }

    public void setY(double y) {
        this.y = y;
    }

    public double getZ() {
        return z;
    }

    public void setZ(double z) {
        this.z = z;
    }

    public float getYaw() {
        return yaw;
    }

    public void setYaw(float yaw) {
        this.yaw = yaw;
    }

    public float getPitch() {
        return pitch;
    }

    public void setPitch(float pitch) {
        this.pitch = pitch;
    }

    @Override
    public String toString() {
        return "HomePosition{" +
                "x=" + x +
                ", y=" + y +
                ", z=" + z +
                ", yaw=" + yaw +
                ", pitch=" + pitch +
                '}';
    }

    public static HomePosition dummyLocation() {
        return of(Bukkit.getWorlds().get(0).getSpawnLocation());
    }

    public static Location fromHomePosition(World world, HomePosition homePosition) {
        return new Location(world, homePosition.getX(), homePosition.getY(), homePosition.getZ(), homePosition.getYaw(), homePosition.getPitch());
    }

    public static HomePosition of(Location location) {
        return new HomePosition(location.getX(), location.getY(), location.getZ(), location.getYaw(), location.getPitch());
    }
}

File: src/main/kotlin/net/onelitefeather/pandorascluster/land/position/HomePosition.kt
package net.onelitefeather.pandorascluster.land.position

import jakarta.persistence.*
import org.bukkit.Bukkit
import org.bukkit.Location
import org.bukkit.World
import org.hibernate.Hibernate

@Entity
data class HomePosition(

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    val id: Long? = null,

    @Column
    val posX: Double = 1.0,

    @Column
    val posY: Double = 1.0,

    @Column
    val posZ: Double = 1.0,

    @Column
    val yaw: Float = 1.0F,

    @Column
    val pitch: Float = 1.0F
) {


    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other == null || Hibernate.getClass(this) != Hibernate.getClass(other)) return false
        other as HomePosition

        return id == other.id
    }

    override fun hashCode(): Int = javaClass.hashCode()

    @Override
    override fun toString(): String =
        this::class.simpleName +
                "(id = $id , posX = $posX , posY = $posY , posZ = $posZ , " +
                "yaw = $yaw , pitch = $pitch)"

    companion object {

        @JvmStatic
        fun fromHomePosition(world: World, homePosition: HomePosition): Location {
            return Location(
                world,
                homePosition.posX,
                homePosition.posY,
                homePosition.posZ,
                homePosition.yaw,
                homePosition.pitch
            )
        }

        @JvmStatic
        fun of(location: Location): HomePosition =
            HomePosition(null, location.x, location.y, location.z, location.yaw, location.pitch)

        @JvmStatic
        fun dummyLocation(): HomePosition = of(Bukkit.getWorlds()[0].spawnLocation)
    }
}
